Fix JS object conversion in unions

Xanewok opened the pull request:
<!-- Please describe your changes on the following line: -->
Requires safe `Heap::boxed` constructor from https://github.com/servo/rust-mozjs/pull/395 (more info on it is in the PR).

Since unions currently assume that their respective members root themselves and can be stored on heap, I modified the union member object conversion branch to convert to a `RootedTraceableBox<Heap<*mut JSObject>>` (which is the currently generated type for objects in said unions).

I did it only for Unions and not dictionaries, since some dictionaries had bare `*mut JSObject` members - is this a mistake and something that needs further fixing?

Does this need a test, e.g. passing a union with object to a function that returns said object, and comparing the results for equality?

r? @jdm

Generated code with this patch (for `StringOrObject`):
```rust
impl FromJSValConvertible for StringOrObject {
    type Config = ();
    unsafe fn from_jsval(cx: *mut JSContext,
                         value: HandleValue,
                         _option: ())
                         -> Result<ConversionResult<StringOrObject>, ()> {
        if value.get().is_object() {
            match StringOrObject::TryConvertToObject(cx, value) {
                Err(_) => return Err(()),
                Ok(Some(value)) => return Ok(ConversionResult::Success(StringOrObject::Object(value))),
                Ok(None) => (),
            }

        }
        // (...)
    }
}

impl StringOrObject {
    // (...)
    unsafe fn TryConvertToObject(cx: *mut JSContext, value: HandleValue) -> Result<Option<RootedTraceableBox<Heap<*mut JSObject>>>, ()> {
        Ok(Some(RootedTraceableBox::from_box(Heap::boxed(value.get().to_object()))))
    }
}
```

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #17011 (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20265)
<!-- Reviewable:end -->

Xanewok committed:
Use mozjs 0.1.14

Xanewok committed:
Support objects in WebIDL unions

Fixes #17011

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

Xanewok commented:
It seems that Taskcluster failed on linux-tests with...
```
It looks like rustup is not installed. See instructions at https://github.com/servo/servo/#setting-up-your-environment
Build FAILED in 0:00:00
```
Should I disregard that?

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20216) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Yes, you can always disregard taskcluster.

jdm commented:
Oh, this is the same commit I reviewed in the typed array PR, isn't it? I think adding a test here would be a useful improvement.

Xanewok committed:
Add WPT test for WebIDL union conversion containing object values

Xanewok commented:
Added a WPT test where check if returned value is the same as the argument passed to an identity function. (On an unrelated note I think it might be a good idea to separate a 'codegen' or similar directory that'd house custom_auto_rooter.html and codegen_union.html and possibly others related in the future?)

On master the test fails with `assert_equals: expected (object) object "[object Object]" but got (string) "[object Object]"`, where it tries to convert the argument to string, but the test passes with this branch.

Is this test good enough?

Now what's left is to use mozjs 0.1.14 when it's published and fix `*mut JSObject` in dictionaries, in addition to unions.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20288) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
1.14 has finally been published.

Xanewok commented:
Updated mozjs to use published 0.1.14. 

As per https://github.com/servo/servo/pull/20267#discussion_r174274790 I didn't include the changes for additional rooting of objects in dictionaries, although the change is pretty straightforward (https://github.com/Xanewok/servo/commit/6380237799eb558a40916229f7b57406c6e05a7b) and can be brought back easily if needed.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 712812d has been approved by `jdm`

<!-- @bors-servo r=jdm 712812d44191bfb916ceaeeada5dc15c2f0dc83d -->

bors-servo commented:
:hourglass: Testing commit 712812d44191bfb916ceaeeada5dc15c2f0dc83d with merge e68d4f45552ed2fdf02853696b3dbe433eb5b195...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3889)

Xanewok commented:
It seems like this is related to the mozangle build failure?
One of the errors:
```
cargo:warning=gfx/angle/checkout/src/common/angleutils.h:33:5: error: unknown type name 'constexpr'
```
And warnings:
```
cargo:warning=gfx/angle/checkout/include/GLSLANG/ShaderVars.h:19:26: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/include/GLSLANG/ShaderVars.h:48:6: warning: scoped enumerations are a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/include/GLSLANG/ShaderVars.h:289:23: warning: defaulted function definitions are a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/include/GLSLANG/ShaderLang.h:276:26: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/include/GLSLANG/ShaderLang.h:459:18: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/src/common/angleutils.h:34:22: warning: defaulted function definitions are a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/src/common/angleutils.h:37:39: warning: deleted function definitions are a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/src/common/angleutils.h:38:42: warning: deleted function definitions are a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/src/common/angleutils.h:108:10: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/src/common/angleutils.h:108:24: warning: range-based for loop is a C++11 extension [-Wc++11-extensions]
cargo:warning=gfx/angle/checkout/src/common/debug.h:37:21: warning: alias declarations are a C++11 extension [-Wc++11-extensions]
warning: build failed, waiting for other jobs to finish...
```

Xanewok commented:
@bors-servo retry

bors-servo commented:
@Xanewok: :key: Insufficient privileges: and not in try users

jdm commented:
@bors-servo retry
* #20281

Xanewok commented:
Ah, I was confused by the star, my bad!

bors-servo commented:
:hourglass: Testing commit 712812d44191bfb916ceaeeada5dc15c2f0dc83d with merge e597cd9e1adc23ae30587ff6bffc05119ac33fb9...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/11033), [arm32](http://build.servo.org/builders/arm32/builds/11271), [arm64](http://build.servo.org/builders/arm64/builds/11303), [linux-dev](http://build.servo.org/builders/linux-dev/builds/11341), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/8220), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/8235), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/11397), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3791), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3817), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/8064), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/8132), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3893), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3839), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/6732)
Approved by: jdm
Pushing e597cd9e1adc23ae30587ff6bffc05119ac33fb9 to master...

Xanewok commented:
:tada: Now onto the typed arrays!
